### PR TITLE
Expose CloudName for Sovereign Clouds

### LINKED ISSUE #5: Set cloud name
Ability to set cloud name to enable azure provider to work in Sovereign cloud.

### PR DESCRIPTION
**Pull request by sudharsan89**
Attempts to fix #5 . 

I am trying to parse the cloudname and set the environment in Provider struct instead of passing around and parsing the cloudname. If there are disadvantages to this, will go back to how it was.

### CODE REVIEW

**Comment by sudharsan89** [comment]
@ritazh @aramase  PTAL

File: README.md [hunk 1]
```diff
@@ -73,6 +74,7 @@ The Azure Key Vault Provider offers two modes for accessing a Key Vault instance
     | provider       | yes      | specify name of the provider                                    | ""            |
     | usePodIdentity | no       | specify access mode: service principal or pod identity          | "false"       |
     | keyvaultName   | yes      | name of a Key Vault instance                                    | ""            |
+    | cloudName      | no       | Name of the azure cloud based on azure go sdk                   | ""            |
```
Comment (aramase) [inline README.md]: nit: can we add allowed values for cloud names -  `AzurePublicCloud`, `AzureUSGovernmentCloud`, `AzureChinaCloud`, `AzureGermanCloud`

File: provider.go [hunk 2]
```diff
@@ -301,6 +294,13 @@ func (p *Provider) MountSecretsStoreObjectContent(ctx context.Context, attrib ma
 	if tenantID == "" {
 		return fmt.Errorf("tenantId is not set")
 	}
+
+	azureCloudEnv, err := ParseAzureEnvironment(cloudName)
+	if err != nil {
+		log.Infof("cloudname is not valid")
```
Comment (aramase) [inline provider.go]: since we are returning the error here, we can remove this log

File: provider.go [hunk 3]
```diff
@@ -301,6 +294,13 @@ func (p *Provider) MountSecretsStoreObjectContent(ctx context.Context, attrib ma
...
+		return err
```
Comment (aramase) [inline provider.go]: `return fmt.Errorf("cloudName is not valid, error: %v", err)`

File: provider.go [hunk 4]
```diff
@@ -301,6 +294,13 @@ func (p *Provider) MountSecretsStoreObjectContent(ctx context.Context, attrib ma
 	if tenantID == "" {
 		return fmt.Errorf("tenantId is not set")
 	}
+
+	azureCloudEnv, err := ParseAzureEnvironment(cloudName)
```
Comment (aramase) [inline provider.go]: Can we add a unit test for `ParseAzureEnvironment` to check we are getting the azure envs as expected for different cloud names?

File: README.md (hunk 1 above)
Comment (sudharsan89) [reply README.md]: Done

File: provider.go (hunk 2 above)
Comment (sudharsan89) [reply provider.go]: Done

File: provider.go (hunk 3 above)
Comment (sudharsan89) [reply provider.go]: Done

File: provider.go (hunk 4 above)
Comment (sudharsan89) [reply provider.go]: Added unit tests

**Comment by sudharsan89** [comment]
> Can we add the optional `cloudName` in the example/* manifests similar to [this](https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/examples/nginx-pod-secrets-store-inline-volume-pod-identity.yaml#L36).

Ah, will address this

**Comment by sudharsan89** [comment]
@aramase updated the PR, PTAL

File: provider_test.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package main
+
+import (
+	"testing"
+	"strings"
+)
+
+func TestParseAzureEnvironment(t *testing.T) {
+	envNamesArray := []string{"AZURECHINACLOUD", "AZUREGERMANCLOUD", "AZUREPUBLICCLOUD", "AZUREUSGOVERNMENTCLOUD"}
```
Comment (ritazh) [inline provider_test.go]: Can you add a test case for `""`, which should equal to `azure.PublicCloud`

**Comment by ritazh** [comment]
Thanks for the PR @sudharsan89 🎉

File: provider_test.go (hunk 5 above)
Comment (sudharsan89) [reply provider_test.go]: Will fix this asap and also do some testing on the PR.

**Comment by ritazh** [comment]
@sudharsan89 _bump_ pls let us know if you need any help on this PR

**Comment by aramase** [comment]
@sudharsan89 Did you get a chance to look at the comments? Let me know if I can help with the PR if possible.

**Comment by sudharsan89** [comment]
> @sudharsan89 Did you get a chance to look at the comments? Let me know if I can help with the PR if possible.

Will take a look sometime this week

File: provider_test.go (hunk 5 above)
Comment (sudharsan89) [reply provider_test.go]: Done, added the test case. It is moved to a new file

**Comment by sudharsan89** [comment]
@ritazh @aramase PTAL

File: pkg/azure/provider.go [hunk 6]
```diff
@@ -277,6 +275,12 @@ func (p *Provider) MountSecretsStoreObjectContent(ctx context.Context, attrib ma
 	if tenantID == "" {
 		return fmt.Errorf("tenantId is not set")
 	}
+
+	azureCloudEnv, err := ParseAzureEnvironment(cloudName)
+	if err != nil {
+		return fmt.Errorf("cloudName is not valid, error: %v", err)
```
[the remark this answers is not in the document]
Comment (sudharsan89) [reply pkg/azure/provider.go]: Done

File: README.md [hunk 7]
```diff
@@ -80,6 +81,8 @@ Create a `secretproviderclasses` resource to provide provider-specific parameter
     | provider       | yes      | specify name of the provider                                    | ""            |
     | usePodIdentity | no       | specify access mode: service principal or pod identity          | "false"       |
     | keyvaultName   | yes      | name of a Key Vault instance                                    | ""            |
+    | cloudName      | no       | Name of the azure cloud based on azure go sdk (AzurePublicCloud,| ""            |
```
[the remark this answers is not in the document]
Comment (sudharsan89) [reply README.md]: Done

File: README.md [hunk 8]
```diff
@@ -57,6 +57,7 @@ Create a `secretproviderclasses` resource to provide provider-specific parameter
       parameters:
         usePodIdentity: "false"         # [OPTIONAL for Azure] if not provided, will default to "false"
         keyvaultName: "kvname"          # the name of the KeyVault
+        cloudName: "cloudname"          # [OPTIONAL for Azure] if not provided, azure environment will default to AzurePublic Cloud
```
[the remark this answers is not in the document]
Comment (sudharsan89) [reply README.md]: Done

**Comment by sudharsan89** [comment]
@aramase PTAL

**Pull request merged**